pdns/misc.cc

```diff
@@ -225,6 +227,39 @@ auto pdns::getMessageFromErrno(const int errnum) -> std::string
   return message;
 }
 
+auto pdns::OpenSSL::error(const std::string& errorMessage) -> std::runtime_error
+{
+  unsigned long errorCode = 0;
+  auto fullErrorMessage{errorMessage};
+#if OPENSSL_VERSION_MAJOR >= 3
+  const char* filename = nullptr;
+  const char* functionName = nullptr;
+  int lineNumber = 0;
+  while ((errorCode = ERR_get_error_all(&filename, &lineNumber, &functionName, nullptr, nullptr)) != 0) {
+    fullErrorMessage += std::string(": ") + ERR_error_string(errorCode, nullptr);
```

Comment:
```
    ERR_error_string() generates a human-readable string representing the
     error code e and places it in buf.  buf must be at least 256 bytes long.
     If buf is NULL, the error string is placed in a static buffer.  Note that
     this function is not thread-safe and does no checks on the size of the
     buffer; use ERR_error_string_n() instead.
```

Reply:
I prefer a thread safe solution.

Reply:
I opted for the following because `ERR_error_string_n()` doesn't really tell me how to find out the required buffer length.

```
    const auto* lib = ERR_lib_error_string(errorCode);
    const auto* reason = ERR_reason_error_string(errorCode);
    fullErrorMessage += std::string(": ") + std::to_string(errorCode) + ":" + lib + "::" + reason;
```

Reply:
The !OpenSSL 3 code still uses `ERR_error_string`

Reply:
Ah... the editor shows it as comments and it's not easy to notice.